utilities/transactions/write_unprepared_txn.cc

```diff
@@ -139,27 +146,34 @@ Status WriteUnpreparedTxn::SingleDelete(ColumnFamilyHandle* column_family,
 }
 
 Status WriteUnpreparedTxn::MaybeFlushWriteBatchToDB() {
+  const bool kPrepared = true;
+  Status s;
+
+  bool needs_mark = log_number_ == 0;
+
   if (max_write_batch_size_ != 0 &&
       write_batch_.GetDataSize() > max_write_batch_size_) {
     assert(GetState() != PREPARED);
-    const bool kPrepared = false;
-    return FlushWriteBatchToDB(kPrepared);
+    s = FlushWriteBatchToDB(!kPrepared);
+    if (s.ok()) {
+      assert(log_number_ > 0);
+      // This is done to prevent WAL files after log_number_ from being
+      // deleted, because they could potentially contain unprepared batches.
+      if (needs_mark) {
+        dbimpl_->logs_with_prep_tracker()->MarkLogAsContainingPrepSection(
+            log_number_);
+      }
+    }
   }
-  return Status::OK();
+  return s;
 }
 
 void WriteUnpreparedTxn::UpdateWriteKeySet(uint32_t cfid, const Slice& key) {
-  if (write_set_keys_.count(cfid) == 0) {
-    auto comp_map = *wupt_db_->GetCFComparatorMap();
-    auto cmp = comp_map[cfid];
-    write_set_keys_[cfid] = CFKeys(SetComparator(cmp));
-  }
-
-  write_set_keys_[cfid].insert(key.ToString());
+  write_set_keys_[cfid].push_back(key.ToString());
```

Comment:
I do not want to keep the PR any longer so can you add a TODO to replace std::vector with a std::string? Then we can do a simple memcpy and avoid the overhead of creating the string object per key.